### PR TITLE
Support compliant coap-proxy.

### PR DESCRIPTION
**Pull request by boaks**
Supports option proxy-uri, proxy-scheme, and coap-uri.
Move proxy-message-deliverer to top-level class in library module.
Deprecate static CoapTranslator in order to replace it with a modifiable
Coap2CoapTranslator. Restore old proxy-resources for backwards
compatibility and deprecate them.
Introduce the new proxy2 resources with cleaned up execution and new
functionality.

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

**Comment by boaks** [comment]
I get more and more the impression, that it doesn't make too much sense to try to migrate the old proxy into a compliant and extendable version.
Currently my preference is to revert the proxy module to version 2.0 and start a second proxy2 module.

**Comment by boaks** [comment]
> Is the old proxy-module "buggy" 

I would not say, that it is buggy; It sticks more to a old spec. with the option proxy-url as mandatory. And it's pretty hard to adapt it for own "special reverse proxies".
According some questions on stackoverflow, some tests with classic http-proxies, and some answers from the core-list, I started therefore to reimplement it, also supporting the other flavours with "proxy-scheme" or just an other destination in "uri-host/uri-port".
But both, the other flavour and the extendability, is hard to achieve with the 2.0 implementation, because that uses a lot of "static" methods.

File: californium-proxy/src/main/java/org/eclipse/californium/proxy/Coap2CoapTranslator.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+import org.eclipse.californium.core.coap.CoAP.ResponseCode;
+import org.eclipse.californium.core.coap.CoAP.Type;
+import org.eclipse.californium.core.coap.OptionSet;
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.core.coap.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Static class that provides the translations between the messages from the
```
[the remark this answers is not in the document]
Comment (boaks) [reply californium-proxy/src/main/java/org/eclipse/californium/proxy/Coap2CoapTranslator.java]: sure

File: californium-proxy/src/main/java/org/eclipse/californium/proxy/CoapUriTranslator.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+import org.eclipse.californium.core.coap.CoAP.ResponseCode;
+import org.eclipse.californium.core.coap.OptionSet;
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.elements.util.StringUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Static class that provides the translations between the messages from the
```
[the remark this answers is not in the document]
Comment (boaks) [reply californium-proxy/src/main/java/org/eclipse/californium/proxy/CoapUriTranslator.java]: also sure.

**Comment by sbernard31** [comment]
> It sticks more to a old spec.

An old obsolete one ? or this is just an alternative spec ?

The new implementations allow to do the same thing as the old one but is more flexible ?
Or both implementation does 2 different things ?

Some other solutions : 
 - do not release the new proxy stuff in 2.1 ?
 - or do not release a 2.1 and directly go to 3.0 ?

(the 4 solutions sounds OK to me)

**Comment by boaks** [comment]
The proxy-uri is still valid, but the other flavours are added and so in many cases will be the preferred flavour. proxy-uri will only be requried, if the uri doesn't fir into a coap-uri.

> The new implementations allow to do the same thing as the old one but is more flexible ?

Yes, but the API will differ, and in my impression too much, that the approach to upgrade the proxy "in place" would have benefits.

My current plan about the release 2.1:
Release 2.1 independently from the new proxy (e.g. new proxy as experimental with currently no API grants). 

(But I have currently too many other tasks, so I guess, I have to postpone to do the pending tasks for 2.1 to next week.)

**Comment by boaks** [comment]
See PR #1208

**Pull request closed without merging**